### ISSUE TITLE
[Subscriptions][Stock] If I edit a subscription while open, I can go into negative stock levels

### ISSUE THREAD

**Issue by sstead** [openfoodfoundation/openfoodnetwork#2505]
## Description
I'm a shop manager and I have an open order cycle with open subscription orders attached. A product is on zero stock. I edit one of the open susbcriptions, to add this item to the subscription. The stock level goes into negative values, even though my instance is set to not allow backorders. The sub order will be confirmed with this product in it at the close of the OC.

## Expected Behavior
The item should be added to the susbcription, but there's insufficient stock, so the customer should get 0 QTY.

## Actual Behavior
Rather than the customer getting 0 QTY of the product that was added to their subscription, they get the amount that was added, but the stock level goes into -ve values.

## Steps to Reproduce
1. Setup a shedule OC with an open subsccription in it.
2. Set the on hand value of a product that's not in the susbcription to zero. 
3. Now edit this susbcription and add the product with zero stock to it.
4. See if the stock level goes to -ve.
5. You'll see the order get confirmed as though there was enough stock on hand, when really the order should have 0 QTY for that item.

## Animated Gif/Screenshot
https://www.useloom.com/share/b164f07ce7984eaf8cdd09c2c725eb1c

## Context
This could arise when subs is used broadly. We need to be confident that the stock control functionality isn't compromised because it's a central OFN feature.

## Severity
s2 

## Your Environment
Staging1
PC Chrome

## Possible Fix
Ensure that stock level is checked when a sub is edited, if stock level is zero, the sub order should have 0 QTY and insufficient stock messaging.

**Comment by daniellemoorhead** [comment]
@sstead this is listed as a severity 2 bug yet it hasn't come up in our conversations about what needs to be completed for subscriptions to be "done" for now.

Is this actually an s2? Is it needed? Any s2 bugs are supposed to be picked up as a priority...yet this one seems to have slipped through without being noticed....

What should we do with it?

**Comment by sstead** [comment]
Eeep sorry. Yes this needs fixing, before subs is used more widely. Its not super likely to occur (maybe a case for s3) but it’s a bug in a core feature (stock control) so could also be s2?

**Comment by daniellemoorhead** [comment]
@kirstenalarsen @myriamboure @RachL what do you think? Worthy of an s2? Should be fixed ASAP?

**Comment by kirstenalarsen** [comment]
i think s3 - not catastrophic if it does happen, they can workaround / amend order, likely only a few - I'd say s3

**Comment by kristinalim** [comment]
Estimated time: 2 hours

This seems related to #2507. In addition to the time needed for that, estimating 2 hours for this.

**Comment by luisramos0** [comment]
I am a bit confused with this one, it looks like a feature in the code, i mean, the stock verification is explicitly skipped when line items are added (and also when quantities are added)...
I dont know subs very well... why are we skipping stock verification in subs?
@kristinalim any clue? shall I just enable stock verification in this page?

the code (skip_stock_check: true):
https://github.com/openfoodfoundation/openfoodnetwork/blob/568e3003ba43c713ac4fb5a57fddfbca086f7b98/app/services/line_item_syncer.rb#L31-L34

**Comment by kristinalim** [comment]
@luisramos0 Stock validation is handled upon subscription _placement_. See here: https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/jobs/subscription_placement_job.rb#L35 The subscription code does not really simply validate (reject/accept), but it caps the order quantity to the maximum that is possible.

So I think the issue is:

* `LineItemSyncer` skips stock validation for both creation and update of subscriptions.
* Placement of order is supposed to handle capping of quantity in orders.
* But when updating subscriptions with existing orders, we don't need to place orders anymore. So the line item quantities are not capped.

It seems like enabling stock verification for subscription _updates_ would solve this issue. Although this will require some shuffling of code among the subscription-related services...

**Comment by luisramos0** [comment]
Excellent @kristinalim thanks for the explanation!

When we update a subscription, the OrderSyncer is called and it calls the line_item_syncer (and we do everything with skip_stock_check). SubscriptionPlacement does the order completion. That makes sense!

I dont understand why order_syncer method **future_and_undated_orders** has this name, this is also returning orders from the current OC that were already placed... and the problem is only with these orders. When we update a subscription we dont need to check stock for future orders, only orders already completed in the current OC that will have their stock updated.

So, following your suggestion I believe I need to change OrderSyncer to process completed orders differently and cap quantities to stock levels (which will include using some code from OrderPlacementJob). Is this what you meant?

**Comment by kristinalim** [comment]
@luisramos0 Yes, that sounds right to me! :slightly_smiling_face:

**Comment by luisramos0** [comment]
all right, product question:

I think this will be bit tricky but needs to be done...

As a manager, when I edit an open subscription to add a product that is out of stock, the subscription orders in the current order cycle are completed already and we are going to add this product. The "expected behaviour" above says "so the customer should get 0 QTY.". That's ok :+1:
1. It will be a bit more work but I guess we need to use existing quantity even if not enough for the full order? if there are 3 in stock and added quantity is 5, we should still add the 3 to the order? (that's what happens in the normal subscription placement process). Right?
2. Also, should we send email to customer informing of the change?
When you change a subscription in the future, the orders were not placed yet, so there is no need to send email to customer. In this case, I think we need to inform the customer.
3. I think we also need to send the manager an email saying what happened (out of stock or insufficient stock).

Basically, if the change involves an order already completed (the order in the current order cycle) we need to repeat the placement job: cap quantity to available stock, send email to customer and manager with changes.

I am a bit new to subs... does this make any sense?

**Comment by luisramos0** [comment]
the easier option (already done in associated PR): if insufficient stock, we just don't add the item to the order. That will not require emails not capping logic... this is at least one day of dev less and also less testing.

Now that I write this, sounds like an acceptable shortcut to the more complicated solution above with emails and stock capping. What do you think?

**Comment by luisramos0** [comment]
I have gone for the simple solution, I dont think we need to send emails in this case and I dont thing it's a big deal if we fail without capping stock (add available quantity to the order).
